src/app/tests/TestICDManager.cpp

```diff
@@ -15,17 +15,204 @@
+     */
+    static void AdvanceClockAndRunEventLoop(uint32_t time_ms)
+    {
+        LayerSocketsLoop & layer = static_cast<LayerSocketsLoop &>(DeviceLayer::SystemLayer());
+        gMockClock.AdvanceMonotonic(System::Clock::Timeout(time_ms));
+
+        layer.PrepareEvents();
+        layer.WaitForEvents();
+        layer.HandleEvents();
+    }
+
+    static void TestICDModeIntervals(nlTestSuite * aSuite, void * aContext)
+    {
+        TestContext * ctx                   = static_cast<TestContext *>(aContext);
+        Clock::ClockBase * const savedClock = &SystemClock();
+
+        // After the init we should be in active mode
+        NL_TEST_ASSERT(aSuite, ctx->mICDManager.mOperationalState == ICDManager::OperationalState::ActiveMode);
+        AdvanceClockAndRunEventLoop(IcdManagementServer::GetInstance().GetActiveModeInterval() + 1);
+
+        // Active mode interval expired, ICDManager transitioned to the IdleMode.
+        NL_TEST_ASSERT(aSuite, ctx->mICDManager.mOperationalState == ICDManager::OperationalState::IdleMode);
+        AdvanceClockAndRunEventLoop(IcdManagementServer::GetInstance().GetIdleModeInterval() + 1);
+        // Idle mode interval expired, ICDManager transitioned to the ActiveMode.
+        NL_TEST_ASSERT(aSuite, ctx->mICDManager.mOperationalState == ICDManager::OperationalState::ActiveMode);
+
+        // Events updating the Operation to Active mode can extend the current active mode time by 1 Active mode threshold.
+        // Kick an active Threshold just before the end of the Active interval and validate that the active mode is extended.
+        AdvanceClockAndRunEventLoop(IcdManagementServer::GetInstance().GetActiveModeInterval() - 1);
+        ctx->mICDManager.UpdateOperationState(ICDManager::OperationalState::ActiveMode);
+
+        AdvanceClockAndRunEventLoop(IcdManagementServer::GetInstance().GetActiveModeThreshold() / 2);
+        ctx->mICDManager.UpdateOperationState(ICDManager::OperationalState::ActiveMode);
+        AdvanceClockAndRunEventLoop(IcdManagementServer::GetInstance().GetActiveModeThreshold());
+        NL_TEST_ASSERT(aSuite, ctx->mICDManager.mOperationalState == ICDManager::OperationalState::IdleMode);
+
+        Clock::Internal::SetSystemClockForTesting(savedClock);
+    }
+
+    static void TestKeepActivemodeRequests(nlTestSuite * aSuite, void * aContext)
+    {
+        TestContext * ctx = static_cast<TestContext *>(aContext);
+
+        Clock::ClockBase * const savedClock = &SystemClock();
+        // Setting a requirement will transition the ICD to active mode.
+        ctx->mICDManager.SetKeepActiveModeRequirements(ICDManager::KeepActiveFlags::kCommissioningWindowOpen, true);
+        NL_TEST_ASSERT(aSuite, ctx->mICDManager.mOperationalState == ICDManager::OperationalState::ActiveMode);
+        // Advance time so active mode interval expires.
+        AdvanceClockAndRunEventLoop(IcdManagementServer::GetInstance().GetActiveModeInterval());
```

Comment:
In previous test we waited GetActiveModeInterval() + 1 to ensure triggering the timer. I know this all boils down to a >= or > somewhere in the SystemTimer but shoudln't we just keep the +1 for consistency? Unless here we specifically need to be exactly on threshold for keep actives?

Reply:
added +1 to keep consistency everywhere